### PR TITLE
Add a label command to kubectl.

### PR DESCRIPTION
**Pull request by brendandburns**
@bgrant0607 @jlowdermilk @smarterclayton

### CODE REVIEW

**Comment by smarterclayton** [comment]
Should this support bulk assignment from day one?  Seems like this is the equiv style command as update.

**Comment by bgrant0607** [comment]
Yes, it should support bulk assignment.

File: docs/kubectl.md [hunk 1]
```diff
@@ -1037,3 +1038,57 @@ Usage:
...
+
+  $ kubectl label --overwrite pods foo status=unhealthy
+  <update a pod with the label 'status' and the value 'unhealthy' overwritting an existing value>
+  
+  $ kubectl label pods foo status=unhealthy --resource-version=1
+  <update a pod with the label 'status' and the value 'unhealthy' if the resource is unchanged from version 1>
+
+Usage:
+```
+  kubectl label [--overwrite] <resource> <name> <key-1>=<val-1> ... <key-n>=<val-n> [--resource-version=<version>] [flags]
```
Comment (bgrant0607) [inline docs/kubectl.md]: We use `<id>` for all other commands. We should be consistent, and convert them all to `<name>` at some point (soon).

File: docs/kubectl.md [hunk 2]
```diff
@@ -390,6 +390,7 @@ Additional help topics:
   kubectl run-container Run a particular image on the cluster.
   kubectl stop          Gracefully shutdown a resource
   kubectl expose        Take a replicated application and expose it as Kubernetes Service
+  kubectl label         Update the labels on a resource
```
Comment (bgrant0607) [inline docs/kubectl.md]: Did you decide not to add commands to remove or change labels? Or do you imagine that would be done via flags?

File: pkg/kubectl/cmd/label.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/meta"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/spf13/cobra"
+)
+
+func (f *Factory) NewCmdLabel(out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "label [--overwrite] <resource> <name> <key-1>=<val-1> ... <key-n>=<val-n> [--resource-version=<version>]",
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/label.go]: A thought for using this to clear labels as well; make the labels to apply a flag and add a flag to clear labels. E.g.
```
label [--overwrite] <resource> <name> --labels=<key-1>=<val-1>... --clear-labels=<key-2>,<key-3>
```

Comment (smarterclayton) [inline pkg/kubectl/cmd/label.go]: What's the most common action?  Adding/overwriting, then clearing?  If adding is sufficiently more prevalent that clearing, it's an argument to make the common action use the arguments.

Comment (bgrant0607) [inline pkg/kubectl/cmd/label.go]: Adding a label or changing a value is most common.

Comment (bgrant0607) [inline pkg/kubectl/cmd/label.go]: I could imagine a flag to remove labels by key.

File: docs/kubectl.md (hunk 2 above)
Comment (smarterclayton) [inline docs/kubectl.md]: At some point we should be worried about verb overload for new users - I think one lesson we took from our previous clients was that as the number of objects increase, having multiple verb forms of a command started to crowd the limited space in cli help.  When dealing with concepts that were common across all objects, having add-X and remove-X commands vs just the single "work with X" command detracted from initial usability.

**Comment by brendandburns** [comment]
I was thinking about a postfix minus sign for delete. eg. ```kubectl label pods foo bar-```  I'd prefer a prefix, but I don't think the flags library can handle that.

File: docs/kubectl.md (hunk 1 above)
Comment (brendandburns) [inline docs/kubectl.md]: ok, not in this PR ;)

File: docs/kubectl.md (hunk 2 above)
Comment (brendandburns) [inline docs/kubectl.md]: Changing labels is done by specifying the --overwrite flag

```kubectl label pods foo baz=bar --overwrite ```

I hadn't thought about remove before, but now I'm thinking about:

```kubectl label pods foo baz-```

File: pkg/kubectl/cmd/label.go (hunk 3 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/label.go]: addressed elsewhere.

**Comment by brendandburns** [comment]
If the remove label syntax is acceptable folks, I'll code that up and send this for review.

**Comment by jlowdermilk** [comment]
> I was thinking about a postfix minus sign for delete. eg. `kubectl label pods foo bar-`
How would that work with removing multiple labels?

File: docs/kubectl.md (hunk 2 above)
Comment (smarterclayton) [inline docs/kubectl.md]: ----- Original Message -----
> > @@ -390,6 +390,7 @@ Additional help topics:
> >    kubectl run-container Run a particular image on the cluster.
> >    kubectl stop          Gracefully shutdown a resource
> >    kubectl expose        Take a replicated application and expose it as
> >    Kubernetes Service
> > +  kubectl label         Update the labels on a resource
> 
> Changing labels is done by specifying the --overwrite flag
> 
> ```kubectl label pods foo baz=bar --overwrite ```
> 
> I hadn't thought about remove before, but now I'm thinking about:
> 
> ```kubectl label pods foo baz-```

I thought of this exact same thing for env vars, and a trailing minus seemed correct.

**Comment by brendandburns** [comment]
@jlowdermilk 
```kubectl label pods foo bar- baz- blah- another=value removethis- add=this ...```
--brendan

**Comment by brendandburns** [comment]
what does bulk update mean in this context?  Do we want to support a label query?  Do we want to support labeling all resources in a collection?

This is kind of unlike update because we don't specify files to relabel.

I'd prefer to add bulk update in a separate PR.

**Comment by jlowdermilk** [comment]
Remove label and bulk update as separate PR sg. LGTM.

**Comment by smarterclayton** [comment]
> On Feb 9, 2015, at 3:46 PM, Brendan Burns <notifications@github.com> wrote:
> 
> what does bulk update mean in this context? Do we want to support a label query? Do we want to support labeling all resources in a collection?
> 
I think selecting a set of things to make more selective (by adding labels) prior to adding a new thing that uses the old label may not be an uncommon use of this command
> This is kind of unlike update because we don't specify files to relabel.
> 
Hrm - it's like delete though, and delete lets you use a directory of config to select the items to delete.
> I'd prefer to add bulk update in a separate PR.
> 
I'm fine with it being a follow on that is to be done before 1.0

**Comment by brendandburns** [comment]
Remove implemented.

@smarterclayton I 100% agree that it will be a common use case.

Promise to do 2 PRs before 1.0 (probably this week):
  * use a selector to select objects, update all objects matched.
  * load objects from a config directory

sg?

File: pkg/kubectl/cmd/label.go (hunk 3 above)
Comment (jlowdermilk) [inline pkg/kubectl/cmd/label.go]: Add remove label arg to usage string.

File: pkg/kubectl/cmd/label.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+		Short: "Update the labels on a resource",
+		Long: `Update the labels on a resource.
+
+If --overwrite is true, then existing labels can be overwritten, otherwise attempting to overwrite a label will result in an error.
+If --resource-version is specified, then updates will use this resource version, otherwise the existing resource-version will be used.
+
+Examples:
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/label.go]: Add remove label example.

**Comment by smarterclayton** [comment]
SGTM

----- Original Message -----

File: docs/kubectl.md (hunk 2 above)
Comment (bgrant0607) [inline docs/kubectl.md]: I'm fine with that, so long as we're consistent with the pattern. We're probably going to need a kubectl-conventions.md.

File: pkg/kubectl/cmd/label.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+			mapping, namespace, name := util.ResourceFromArgs(cmd, res, mapper, cmdNamespace)
+			client, err := f.RESTClient(cmd, mapping)
+			checkErr(err)
+
+			labels, remove, err := parseLabels(args[2:])
+			checkErr(err)
+			overwrite := util.GetFlagBool(cmd, "overwrite")
+			resourceVersion := util.GetFlagString(cmd, "resource-version")
+
+			obj, err := updateObject(client, mapping, namespace, name, func(obj runtime.Object) runtime.Object {
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/label.go]: I didn't see a response to the bulk-update issue.

Example is here:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/cmd/update.go#L57

Obviously, a resource-version precondition would be incompatible with a bulk update.

Comment (bgrant0607) [inline pkg/kubectl/cmd/label.go]: Ok, comment didn't show in file view, but saw it in email. I'm fine with a separate PR.

**Comment by bgrant0607** [comment]
Reassigning to @jlowdermilk since he left comments that still look unresolved.

**Comment by brendandburns** [comment]
PR updated to document the remove label syntax.

**Comment by jlowdermilk** [comment]
lgtm

**Pull request merged**